### ISSUE TITLE
Deploying to npm or unpkg

### ISSUE THREAD

**Issue by pkerpedjiev** [andywer/threads.js#156]
Quick question: do you have a suggestion for how to deploy a library using threads.js?

When building with webpack, two scripts are generated: `mylib.min.js` and `0.mylib.worker.min.js`. If I deploy them as an npm package, I can access `https://unpkg.com/mylib@0.1.0/dist/mylib.min.js` and `https://unpkg.com/mylib@0.1.0/dist/0.mylib.worker.min.js`.

But when I create a test script and include the main script:

```
<script src="https://unpkg.com/mylib@0.1.0/dist/mylib.min.js"></script>
```

It will try to load the worker from `/0.mylib.worker.min.js` and not from its location on unpkg. Is there a way to tell it to look for the worker in the same directory as the main script or do I need to hack this in using one of the techniques described on [stackoverflow](https://stackoverflow.com/questions/2976651/javascript-how-do-i-get-the-url-of-script-being-called)?

## Edit: Final solution

Add an option `_baseURL` (string) to the worker options. It's only picked up by web workers (not in node.js) and let's you prefix the URL with some static base URL.

```js
new Worker("./my-worker", {
  _baseURL: "https://unpkg.com/mylibrary@v1.2.3/"
})
```

**Comment by andywer** [comment]
Hey @pkerpedjiev, that's a pretty good question!

Have you tried to pass a fully-qualified URL to the worker constructor? I imagine something like this:

```js
spawn(new Worker(process.env.NODE_ENV === "production" ? "https://unpkg.com/mylib@0.1.0/dist/0.mylib.worker.min.js" : "./path/to/worker"))
```

Might require a change in the `threads-plugin`, though.

**Comment by pkerpedjiev** [comment]
Hey, thanks for the quick reply! I haven't tried anything beyond what's in the original question. I don't think `process.env.NODE_ENV` will be defined when the script is being served through a regular server. 

An alternative I'm considering is adding a `sed` step in `npm run build` that replaces `./mylib.worker.js` with `getScriptUrl()`.

**Comment by andywer** [comment]
Ahh, sorry, I didn't share that thought:

I suppose you build the library with webpack and publish the build files only. You can set the `NODE_ENV` to `production` in your library's webpack config.

On a second thought, maybe you don't want to webpack-build the package... Do you run webpack and publish the build files or do you intend to ship the source files as a library and let the application that uses your library build it with webpack?

**Comment by pkerpedjiev** [comment]
Yes, I'm using webpack and would build this as a package and publish it. For the time being I don't expect people to use it as an import in a larger application.

I was just confused because the `process.env.NODE_ENV` is in the library source code (rather than webpack.config) so I figured it would be evaluated at runtime rather than build time.

**Comment by andywer** [comment]
To be clear: The `NODE_ENV` check would be in the library source code, but if you build the package with webpack, then you can just use the `DefinePlugin` to statically replace `NODE_ENV` with `production` when building your package, so that you can run tests locally in your package and it will spawn the latest local worker, not the one from unpkg, but when built it will always spawn the unpkg worker.

That was my idea, at least 😉

**Comment by pkerpedjiev** [comment]
Just tried out your suggestion. I got the following error:

```
ReferenceError: falseundefined__webpack__worker__25 is not defined
```

**Comment by pkerpedjiev** [comment]
Just to clarify, I get the error above when trying to run using the dev server:

```
Plugin track pileup failed to instantiate. ReferenceError: falseundefined__webpack__worker__1 is not defined
    at new BAMDataFetcher (bam-fetcher.js:64)
    ...
```

The production build works fine. Here's is the relevant line in the output js file:

```
    this.worker = Object(dist["spawn"])(new dist["Worker"]( true ? getThisScriptLocation() + '/0.higlass-pileup.min.worker.js' : undefined));
```

`getThisScriptLocation()` is a function I added that returns base path where the calling script lives. I looked at `threads-plugin` but wasn't entirely sure where to even begin.

**Comment by andywer** [comment]
That's a bit strange. When I try, my production build fails already, but with a different error.

Code: https://github.com/andywer/threads.js/commit/81e64104e79e6591c7864e5f4b2992ab6bc5e941 ([`feature/support-remote-url-worker`](https://github.com/andywer/threads.js/tree/feature/support-remote-url-worker))
Error: <https://travis-ci.org/andywer/threads.js/jobs/582091315>

**Comment by pkerpedjiev** [comment]
Oh, you know what? You're right. It worked for me because there was already a worker script in the dist directory from a previous build. When I clear it out and run `npm run build` a new one doesn't get generated with the `... process.env.NODE_ENV === "production" ...` line.

**Comment by andywer** [comment]
Also on a second thought, it's not wise to have the ternary operator inside the `new Worker()` as it's not statically analyzable and will throw the `threads-plugin` off its course.

Instead it should rather be something like this to keep it analyzable:

```diff
- spawn(new Worker(process.env.NODE_ENV === "production" ? "https://unpkg.com/mylib@0.1.0/dist/0.mylib.worker.min.js" : "./path/to/worker"))
+ spawn(process.env.NODE_ENV === "production" ? new Worker("https://unpkg.com/mylib@0.1.0/dist/0.mylib.worker.min.js") : new Worker("./path/to/worker"))
```

**Comment by pkerpedjiev** [comment]
I guess you already know this but that works in develop mode but not in production.

Perhaps there can be an option to `Worker` that can take a function which returns prefix for where the worker will be loaded from? Because the current proposal assumes that we know that the generated script will be `0...worker.min.js`. I presume that if there's more than one worker it will be `1...worker.ming.js` and `2.`, etc..

**Comment by andywer** [comment]
I did some stuff 😉

Can you try again with `threads@1.0.0-beta.5-feature-156` and `threads-plugin@1.2.0`?

**Comment by pkerpedjiev** [comment]
Cool! I installed those new packages but If I do what you suggested earlier:

```
    this.worker = spawn(
      process.env.NODE_ENV === 'production' ?
        new Worker(
          `${getThisScriptLocation()}/0.higlass-pileup.worker.min.js`,
        ) : new Worker('./bam-fetcher-worker.js'),
    );
```
Then running `npm run build` doesn't generate `dist/0.higlass-pileup-worker.min.js`. Also, ideally I wouldn't want to add `0.higlass-pileup.worker.min.js` inside my code since that filename is automatically generated by threads.js. Does that make sense?

I don't know how feasible it is but a potential API could look like the following. `threads-plugin` would replace the `'./bam-fetcher-worker.js'` with the generated worker file (e.g. `'0.higlass-pileup-worker.min.js'`). 

```
    this.worker = spawn(
          new ThreadsWorker('./bam-fetcher-worker.js', getThisScriptLocation())
    );
```

Or maybe that's already possible and I'm not calling the API correctly?

**Comment by andywer** [comment]
I think the issue is that the first `new` expression is again not statically analyzable. Pass a string literal like `new Worker("https://unpkg.com/...")` and it should work. Not sure if referencing an env var in the expression or an import from package.json, for instance, would still work.

Got to think about the API change...

**Comment by pkerpedjiev** [comment]
Hmmm, I don't know about hardcoding an unpkg location directly in the source. That seems like it introduces the opposite problem. Instead of being tied to `/script.worker.js` it's now tied to `unpkg.com/.../0.script.worker.js` with the downside that we now need to know the output name of the worker script (e.g. `0...worker`...).

I appreciate your help and sorry for being so particular.

**Comment by pkerpedjiev** [comment]
Just tried my hand at implementing this functionality: https://github.com/andywer/threads.js/pull/161

Seems to work for my use case:

```
    this.worker = spawn(
      new Worker('./bam-fetcher-worker.js',
        {},
        () => getThisScriptLocation()),
    );
```

**Comment by andywer** [comment]
I see your point and appreciate your dedication, we just need a different API change. Adding a new parameter to the constructor for one edge-case that only makes real sense in the browser is a bit much.

I quickly extended the worker options (2nd argument to constructor) by a new property `_baseURL` in the feature branch mentioned before.

Usage looks like this:

```js
new Worker("./path/to/worker", { _baseURL: "https://unpkg.com/..." })
```

Published as `threads@1.0.0-beta.5-feature-156.1`. Can you try that, too, and tell me what you think about this approach?

**Comment by pkerpedjiev** [comment]
That looks great! I just did a quick check and it looks like it works with the caveat that `_baseURL` needs to end in a slash.

I'll test it more extensively this evening. Thanks for your help!!

**Comment by andywer** [comment]
Nice! Are you sure about the trailing slash, though? Since it uses the `URL` constructor it should work either way.

**Comment by pkerpedjiev** [comment]
Yup, this works and finds the script at `http://localhost/dist/0.script.js`:

```
    this.worker = spawn(
      new Worker('./bam-fetcher-worker.js', {
        _baseURL: `${thisScriptLocation}/`,
      }),
    );
```

This does not work. It looks for the script at `http://localhost/0.script.worker.js` when it's actually at `http://localhost/dist/0.script.worker.js`.

```
    this.worker = spawn(
      new Worker('./bam-fetcher-worker.js', {
        _baseURL: `${thisScriptLocation}/`,
      }),
    );
```

**Comment by andywer** [comment]
Ahh, interesting. It's path override vs override host, but keep path. Yet another edge case... 🙈😅

I am happy you could figure it out, though! Shall I merge the PR and release or is there anything left to do?

**Comment by pkerpedjiev** [comment]
I think it looks good. Thanks for all your help!

**Comment by andywer** [comment]
Feature published as `threads@1.0.0-beta.6` 🚀
